libs/utils/analysis/latency_analysis.py

```diff
@@ -0,0 +1,361 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (C) 2015, ARM Limited and contributors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+""" Latency Analysis Module """
+
+import matplotlib.gridspec as gridspec
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import pylab as pl
+import re
+
+from collections import namedtuple
+from analysis_module import AnalysisModule
+from devlib.utils.misc import memoized
+from trappy.utils import listify
+
+# Configure logging
+import logging
+
+# Tuple representing all IDs data of a Task
+TaskData = namedtuple('TaskData', ['pid', 'names', 'label'])
+
+class LatencyAnalysis(AnalysisModule):
+    """
+    Support for plotting Latency Analysis data
+
+    :param trace: input Trace object
+    :type trace: :mod:`libs.utils.Trace`
+    """
+
+    def __init__(self, trace):
+        super(LatencyAnalysis, self).__init__(trace)
+
+###############################################################################
+# DataFrame Getter Methods
+###############################################################################
+
+    @memoized
+    def _dfg_latency_df(self, task):
+
+        if not self._trace.hasEvents('sched_wakeup'):
+            logging.warn('Events [sched_wakeup] not found, '
+                         'cannot compute CPU active signal!')
+            return None
+        if not self._trace.hasEvents('sched_switch'):
+            logging.warn('Events [sched_switch] not found, '
+                         'cannot compute CPU active signal!')
+            return None
+
+        # Get task data
+        td = self._getTaskData(task)
+        if not td:
+            return None
+
+        wk_df = self._dfg_trace_event('sched_wakeup')
+        sw_df = self._dfg_trace_event('sched_switch')
+
+        # Filter Task's WAKEUP events
+        task_wakeup = wk_df[wk_df.pid == td.pid][['success', 'target_cpu', 'pid']]
+
+        # Filter Task's START events
+        task_events = (sw_df.prev_pid == td.pid) | (sw_df.next_pid == td.pid)
+        task_switches_df = sw_df[task_events]\
+            [['__cpu', 'prev_pid', 'next_pid', 'prev_state']]
+
+        # Unset prev_state for switch_in enents, i.e.
+        # we don't care about the status of a task we are replacing
+        task_switches_df.prev_state = task_switches_df.apply(
+            lambda r : np.nan if r['prev_pid'] != td.pid else r['prev_state'], axis = 1)
```

Comment:
To avoid the issue with having numbers as current states, we could do something like this:

```python
def task_state_to_str(state):
    TASK_STATES = {
        0: "R",
        1: "S",
        2: "D",
        4: "T",
        8: "t",
        16: "Z",
        32: "X",
        64: "x",
        128: "K",
        256: "W",
        512: "P"
    }
    TASK_MAX_STATE = 1024

    res = "R"
    if state & (TASK_MAX_STATE - 1) != 0:
        res = ""
    for key in TASK_STATES.keys():
        if key & state:
            res += TASK_STATES[key]
    if state & TASK_MAX_STATE:
        res += "+"
    else:
        res = '|'.join(res)
    return res

task_switches_df.prev_state = task_switches_df.apply(
    lambda r: numpy.nan if r['prev_pid'] != td.pid
                        else task_state_to_str(int(r['prev_state'])),
    axis=1
)
```